File: Graph/827.making-a-large-island.go
package leetcode

// @leet start
type s8 struct {
	grid   [][]int
	areaMP map[int]int
	dirs   []p8
	ret    int
}

func (s *s8) set(pt p8, v int) {
	s.grid[pt.r][pt.c] = v
}

func (s *s8) get(pt p8) int {
	return s.grid[pt.r][pt.c]
}

func (s *s8) dfs(p p8, id int) {
	if !s.inside(p) || s.get(p) != 1 {
		return
	}
	s.areaMP[id]++
	s.set(p, id)
	for _, dir := range s.dirs {
		s.dfs(p.move(dir), id)
	}
}

func (s *s8) find(p p8) {
	total := 0
	seen := map[int]bool{}
	for _, dir := range s.dirs {
		next := p.move(dir)
		if !s.inside(next) {
			continue
		}
		id := s.get(next)
		if seen[id] {
			continue
		}
		total += s.areaMP[id]
		seen[id] = true
	}
	s.ret = max(s.ret, total+1)
}

func (s *s8) inside(pt p8) bool {
	nr, nc := len(s.grid), len(s.grid[0])
	return pt.r >= 0 && pt.r < nr && pt.c >= 0 && pt.c < nc
}

func largestIsland(grid [][]int) int {
	sol := s8{
		areaMP: map[int]int{},
		grid:   grid,
		dirs:   dirs,
	}
	id := 2
	for r := range len(grid) {
		for c := range len(grid[0]) {
			curP := p8{r, c}
			if sol.get(curP) == 1 {
				// sol.curArea = 0
				sol.dfs(curP, id)
				// sol.areaMP[id] = sol.curArea
				id++
			}
		}
	}
	found0 := false
	// find max area
	for r := range len(grid) {
		for c := range len(grid[0]) {
			curP := p8{r, c}
			if sol.get(curP) == 0 {
				found0 = true
				sol.find(curP)
			}
		}
	}

	if !found0 {
		return len(grid) * len(grid[0])
	}
	return sol.ret
}

var dirs = []p8{
	{0, 1},
	{0, -1},
	{-1, 0},
	{1, 0},
}

type p8 struct {
	r, c int
}

func (p p8) move(p1 p8) p8 {
	return p8{p.r + p1.r, p.c + p1.c}
}

// @leet end

File: DFS/827.making-a-large-island.py
from typing import List, Self

# @leet start
from dataclasses import dataclass


@dataclass(slots=True)
class pt:
    r: int
    c: int

    def move(self, d: Self):
        return pt(self.r + d.r, self.c + d.c)


class Solution:
    dirs = [
        pt(0, 1),
        pt(0, -1),
        pt(-1, 0),
        pt(1, 0),
    ]

    def largestIsland(self, grid: List[List[int]]) -> int:
        group_area: dict[int, int] = {0: 0}
        res = 0
        nr, nc = len(grid), len(grid[0])
        group = 2

        def get(p: pt):
            return grid[p.r][p.c]

        def inside(p: pt):
            return p.r >= 0 and p.r < nr and p.c >= 0 and p.c < nc

        def dfs(p: pt, g: int) -> int:
            if not inside(p) or get(p) != 1:
                return 0
            grid[p.r][p.c] = g
            area = 1
            for d in self.dirs:
                area += dfs(p.move(d), g)
            return area

        def findMax(p: pt):
            seen = set()
            cur = 1
            for d in self.dirs:
                ap = p.move(d)
                if inside(ap) and get(ap) not in seen:
                    cur += group_area[get(ap)]
                    seen.add(get(ap))
            return cur

        for r in range(nr):
            for c in range(nc):
                if grid[r][c] == 1:
                    area = dfs(pt(r, c), group)
                    res = max(area, res)
                    group_area[group] = area
                    group += 1

        for r in range(nr):
            for c in range(nc):
                if grid[r][c] == 0:
                    res = max(findMax(pt(r, c)), res)
        return res


# @leet end

